include/alpaka/core/CallbackThread.hpp

```diff
@@ -0,0 +1,77 @@
+/* Copyright 2022 Antonio Di Pilato
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+namespace alpaka::core
+{
+    class CallbackThread
+    {
+        using Task = std::packaged_task<void()>;
+
+    public:
+        ~CallbackThread()
+        {
+            {
+                std::unique_lock<std::mutex> lock{m_mutex};
+                m_stop = true;
+            }
+            m_cond.notify_one();
+            if(m_thread.joinable())
+                m_thread.join();
+        }
+        auto submit(Task&& newTask) -> std::future<void>
+        {
+            auto f = newTask.get_future();
+            {
+                std::unique_lock<std::mutex> lock{m_mutex};
+                m_tasks.emplace(std::move(newTask));
+                if(!m_thread.joinable())
+                {
+                    m_thread = std::thread(
+                        [this]
+                        {
+                            Task task;
+                            while(true)
+                            {
+                                {
+                                    std::unique_lock<std::mutex> lock{m_mutex};
+                                    m_cond.wait(lock, [this] { return m_stop || !m_tasks.empty(); });
+
+                                    if(m_stop && m_tasks.empty())
+                                        break;
+
+                                    task = std::move(m_tasks.front());
+                                    m_tasks.pop();
+                                }
+
+                                task();
+                            }
```

Comment:
Looks good to me.

Personally I would find the code a bit easier to understand it it were written without the lambda and the unique_lock block, like
```c++
    Task task;
    while(true)
    { 
        std::unique_lock<std::mutex> lock{m_mutex};
        while (not m_stop and m_tasks.empty()) {
            m_cond.wait(lock);
        } 
            
        if (m_stop and m_tasks.empty())
            break;
                
        task = std::move(m_tasks.front());
        m_tasks.pop();

        // run the task without holding the lock
        lock.unlock();
        task();
    }   
```
but I suspect it's just a matter of taste.
No need to change it.

Reply:
I followed the same style that is adopted somewhere else in alpaka, i.e. [here](https://github.com/alpaka-group/alpaka/blob/2fc0f04f45a7ea0a7e4ae7d3a5bfd1f031f56fb0/include/alpaka/core/ConcurrentExecPool.hpp#L479-L484), but I agree that the `unlock` method is easier to read.

Reply:
I prefer the implementation from @tonydp03 because it is moving the handling of false wakeups to the std implementation: 
https://en.cppreference.com/w/cpp/thread/condition_variable/wait is pointing out that `cond.wait(mutex,prediction) is equal to what @fwyzard wrote.

Reply:
Please keep the lambda in `m_cond.wait(lock, [this] { return m_stop || !m_tasks.empty(); });`. It is necessary to protect against spurious wakeups.

I would prefer the explicit block to calling `unlock()` manually, but also a matter of taste.

Reply:
> It is necessary to protect against spurious wakeups.

The loop and the lambda are equivalent for that purpose (replacing the loop with an if would indeed be unsafe).
I personally find the loop more readable (one less `not` to keep in mind), but as I wrote it was just a mutter of understanding the code.

Does the fact that the lambda captures `this` (by value) make the code less optimal than the loop ?